### PR TITLE
Permit [eip-2612] support for AMPL

### PR DESCRIPTION
**Pull request by nithinkrishna**
Gasless approvals with permit. 
https://eips.ethereum.org/EIPS/eip-2612

Reference implementations
* Openzeppelin: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/drafts/ERC20Permit.sol
* AToken: https://github.com/ampleforth/protocol-v2/blob/master/contracts/protocol/tokenization/AToken.sol

There's one major discussion point. 

As per the standard openzeppelin implementation, the domain separator is computed in the [contract's constructor](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/drafts/EIP712.sol#L46). However since we have a live contract instance, we cant do that.

In the current approach, I've hardcoded the values that go into the domain separator creation based on the production config. (Address, chainId and token name).

If we dont want to hardcode this, we could create a function which we invoke after upgrade which sets the domain separator after reading the contract's store (more general solution).

### CODE REVIEW

File: contracts/UFragments.sol
Comment (thegostep) [inline contracts/UFragments.sol:336]: add revert message?

File: test/unit/UFragments.ts
Comment (thegostep) [inline test/unit/UFragments.ts:5]: Could use ethers for typed data signing: https://docs.ethers.io/v5/api/signer/#Signer-signTypedData

File: contracts/UFragments.sol
Comment (thegostep) [inline contracts/UFragments.sol:190]: set to external?

**Review by thegostep — APPROVED** [review]
lgtm - though was only able to take a cursory look

File: contracts/UFragments.sol
Comment (nithinkrishna) [reply contracts/UFragments.sol:336]: Dont think revert messages are possible in 0.4.24

File: contracts/UFragments.sol
Comment (nithinkrishna) [reply contracts/UFragments.sol:190]: kept it consistent with `balanceOf` and `totalSupply`. Worthwhile changing all of them to `external`?

Comment (thegostep) [reply contracts/UFragments.sol:190]: yeah - it will give a (small) gas saving

File: test/unit/UFragments.ts
Comment (nithinkrishna) [reply test/unit/UFragments.ts]: Done 👍

File: contracts/UFragments.sol
Comment (nithinkrishna) [reply contracts/UFragments.sol:190]: https://github.com/ampleforth/uFragments/issues/183

File: contracts/UFragments.sol
Comment (brandoniles) [inline contracts/UFragments.sol:81]: ` = bytes('1');` to be explicit? This is what AToken does.

Comment (nithinkrishna) [reply contracts/UFragments.sol:81]: that doesn't work in solidity 0.4.24 :/

**Review by brandoniles — APPROVED** [review]
Thanks!

**Comment by nithinkrishna** [comment]
Will wait for @ahnaguib to review this

File: contracts/UFragments.sol
Comment (nithinkrishna) [inline contracts/UFragments.sol:342]: @ahnaguib How OpenZeppelin implements this

https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/drafts/EIP712Upgradeable.sol#L88

https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/drafts/ERC20PermitUpgradeable.sol#L50

File: package.json
Comment (thegostep) [inline package.json:42]: doesn't seem like this import is required anymore

Comment (nithinkrishna) [inline package.json:42]:
```suggestion
```

File: test/utils/signatures.ts
Comment (nithinkrishna) [inline test/utils/signatures.ts:9]:
```suggestion
```

**Review by brandoniles — CHANGES_REQUESTED** [review]
Until we address the hardcoded separator issue.

**Comment by nithinkrishna** [comment]
Holding off on this until we bump up the token solidity compiler version

**Comment by thegostep** [comment]
OZ released Permit contract today: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v3.4.0/contracts/drafts/ERC20Permit.sol

**Comment by nithinkrishna** [comment]
They've had it in drafts for a while now

**Comment by nithinkrishna** [comment]
@brandoniles I've updated to the new implementation which queries the chainId everytime

File: contracts/UFragments.sol
Comment (brandoniles) [inline contracts/UFragments.sol:202]: Does this need to be public?

Also, basic question but--does this being its own function use more gas than it would if it were just in the body of `DOMAIN_SEPARATOR()`?

Comment (nithinkrishna) [reply contracts/UFragments.sol:202]: Hmm. Not necessarily. But if it is public, off-chain services can look up all the attributes to compute the domain_seperator from the token contract's interface.

```
keccak256(
    abi.encode(
        uFragments.EIP712_DOMAIN(),
        keccak256(bytes(uFragments.name())),
        keccak256(bytes(uFragments.EIP712_REVISION())),
        uFragments.getChainId(),
        uFragments.address
    )
)
```

Comment (nithinkrishna) [reply contracts/UFragments.sol]: Makes sense.

**Review by brandoniles — APPROVED** [review]
LGTM, but wait for at least @thegostep 's approval also.

**Review by thegostep — APPROVED** [review]
lgtm

**Pull request merged**
